tests/agent_features/test_configuration.py

```diff
@@ -949,6 +949,69 @@ def test_initialize_developer_mode(section, expect_error, logger):
         assert "CONFIGURATION ERROR" not in logger.caplog.records
 
 
+newrelic_toml_contents = b"""
+[tool.newrelic]
+app_name = "test11"
+monitor_mode = true
```

Comment:
We have various rule configuration options. You can find them [here](https://github.com/newrelic/newrelic-python-agent/blob/main/newrelic/core/config.py#L726). When I add one of them like the example below. It looks like the parsing fails because it can't handle a list of dictionaries.
```suggestion
monitor_mode = true

url_rules = [{ignore = false, match_expression = ".*\\\.(ace|arj|ini|txt|udl|plist|css|gif|ico|jpe?g|js|png|swf|woff|caf|aiff|m4v|mpe?g|mp3|mp4|mov)$"}]

```

Reply:
I added in a fix for this but I was not able to check it in the test. `fetch_config_setting(settings, "url_rules")` always returns `[]`. Grepping for `url_rules` shows:

```
newrelic/common/agent_http.py
587:            "url_rules": [],

newrelic/core/application.py
399:                            "The URL normalization rules for %r are %r.", self._app_name, configuration.url_rules
412:                    self._rules_engine["url"] = RulesEngine(configuration.url_rules)

newrelic/core/config.py
726:_settings.url_rules = []
```

There is no call to `_process_setting()` for url_rules in `core/config.py`.

The following does work to fix the issue but I do not know if this is correct. Plus it might break any existing INI functionality assuming that is working for url_rules.

```diff
diff --git a/newrelic/config.py b/newrelic/config.py
index 73a847ba..11e5268c 100644
--- a/newrelic/config.py
+++ b/newrelic/config.py
@@ -14,6 +14,7 @@
 
 import configparser
 import fnmatch
+import json
 import logging
 import os
 import sys
@@ -309,6 +310,7 @@ def _process_setting(section, option, getter, mapper):
 
 
 def _process_configuration(section):
+    _process_setting(section, "url_rules", "get", json.loads)
     _process_setting(section, "feature_flag", "get", _map_feature_flag)
     _process_setting(section, "app_name", "get", None)
     _process_setting(section, "labels", "get", _map_labels)
diff --git a/tests/agent_features/test_configuration.py b/tests/agent_features/test_configuration.py
index a359a0de..8e83883b 100644
--- a/tests/agent_features/test_configuration.py
+++ b/tests/agent_features/test_configuration.py
@@ -999,6 +999,9 @@ def test_toml_parse_development():
         assert value.enabled is True
         assert value.ignore_classes[0] == "module:name1"
         assert value.ignore_classes[1] == "module:name"
+        value = fetch_config_setting(settings, "url_rules")
+        assert isinstance(value, list)
+        assert value[0]['ignore'] is False
 
 
 def test_toml_parse_production():
 ```

Also, how do I get rid of this error?

```
tests/agent_features/test_configuration.py::test_toml_parse_production ERROR

================================================================================================================================== ERRORS ===================================================================================================================================
_______________________________________________________________________________________________________________ ERROR at setup of test_toml_parse_production ________________________________________________________________________________________________________________

request = <SubRequest 'collector_available_fixture' for <Function test_toml_parse_production>>, collector_agent_registration = <newrelic.api.application.Application object at 0x7f3b7bf023c0>

    @pytest.fixture(scope="function")
    def collector_available_fixture(request, collector_agent_registration):
        application = application_instance()
        active = application.active
>       assert active
E       AssertionError

tests/testing_support/fixtures.py:277: AssertionError
========================================================================================================================== short test summary info ==========================================================================================================================
ERROR tests/agent_features/test_configuration.py::test_toml_parse_production - AssertionError
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================================================================= 140 passed, 1 error in 1.24s ========================================================================================================================
```

Reply:
Either set a valid `NEW_RELIC_LICENSE_KEY` environment variable or set `NEW_RELIC_DEVELOPER_MODE=true` to avoid sending any test data to our real backend.

Reply:
@hmstepanek URL rules aren't a setting available in config files, they're a serverside setting that gets sent to us on connect. You can't manually create them in the INI file.

Any serverside only settings shouldn't be a problem for this PR, as we won't attempt to read that information at all from a config file.

Reply:
Oh I see! You're right-none of these are processed for the config file. Given that, I think this is good as is then!